sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -236,15 +241,22 @@ private[sql] object PartitioningUtils {
   }
 
   /**
-   * Converts a string to a [[Literal]] with automatic type inference.  Currently only supports
-   * [[IntegerType]], [[LongType]], [[DoubleType]], [[DecimalType.SYSTEM_DEFAULT]], and
-   * [[StringType]].
+   * Converts a string to a [[Literal]] with automatic type inference if no field type specified.
+   * Auto inference only supports [[IntegerType]], [[LongType]], [[DoubleType]],
+   * [[DecimalType.SYSTEM_DEFAULT]], and [[StringType]].
    */
   private[sql] def inferPartitionColumnValue(
+      expectedDT: Option[DataType],
```

Comment:
In the master branch, if `typeInference == false`, it means the data type of partition key will be `StringType` by default, otherwise, it's probably will be `IntegerType`, `LongType` etc. depends on the real value the partition key is.